openvdb_points/tools/PointDataGrid.h

```diff
@@ -360,6 +360,67 @@ class PointDataLeafNode : public tree::LeafNode<T, Log2Dim> {
         mAttributeSet->reorderAttributes(replacement);
     }
 
+    /// @brief Swap the underlying attribute set with the given @a attributeSet.
+    /// This leaf will assume ownership of the given attribute set. The new
+    /// attributes must have an equal descriptor. Additionally, if the new
+    /// attributes have different topology, then the voxel offset values must
+    /// be updated to reflect this new topology after a call to this method.
+    void swap(AttributeSet* attributeSet)
+    {
+        if (!attributeSet) {
+            OPENVDB_THROW(ValueError, "Cannot swap with a null attribute set");
+        }
+
+        if (mAttributeSet->descriptor() != attributeSet->descriptor()) {
+            OPENVDB_THROW(ValueError, "Attribute set descriptors are not equal.");
+        }
+
+        mAttributeSet.reset(attributeSet);
+    }
+
+    /// @brief Sets all of the voxel offset values on this leaf, from the given vector
+    /// of @a offsets. If @a updateValueMask is true, then the active value mask will
+    /// also be updated such that voxels with points are marked as active, and empty voxels
+    /// are marked as inactive.
+    void setOffsets(const std::vector<ValueType>& offsets, const bool updateValueMask = true)
+    {
+        if (offsets.size() != LeafNodeType::NUM_VALUES) {
+            OPENVDB_THROW(ValueError, "Offset vector size doesn't match number of voxels.")
+        }
+
+        for (size_t index = 0; index < offsets.size(); ++index) {
+            setOffsetOnly(index, offsets[index]);
+        }
+
+        if (updateValueMask)
+            this->updateValueMask();
+    }
+
+    /// @brief Validates the voxel offset values on this leaf to ensure that they are both
+    /// monotonically increasing, and that they are within the bounds of the attribute
+    /// arrays on this leaf node. Throws an exception if any of these checks fail.
+    void validateOffsets() const
+    {
+        // Ensure all of the offset values are monotonically increasing
+        for (size_t prevIndex = 0, index = 0; index < BaseLeaf::SIZE; prevIndex = index++) {
+            if (this->getValue(prevIndex) > this->getValue(index)) {
+                OPENVDB_THROW(ValueError, "Voxel offset values are not monotonically increasing");
+            }
+        }
+
+        // Ensure the last voxel's offset value matches the size of each attribute array
+        // on this leaf node
+        for (size_t attributeIndex = 0; attributeIndex < mAttributeSet->size(); ++attributeIndex) {
+            AttributeArray* array = mAttributeSet->get(attributeIndex);
+            if (this->getValue(BaseLeaf::SIZE-1) != array->size()) {
+                std::ostringstream message;
+                message << "Last voxel offset value does not match size of attribute array at index ";
+                message << attributeIndex;
+                OPENVDB_THROW(ValueError, message.str());
```

Comment:
Suggest that to avoid the custom message stuff, we do this in two parts:

* check all attribute array sizes match in a similar loop to above (ie no need to say which index isn't correct)
* check the attribute array size matches the leaf offset